README
libbsd - Utility functions from BSD systems

This library provides useful functions commonly found on BSD systems,
and lacking on others like GNU systems, thus making it easier to port
projects with strong BSD origins, without needing to embed the same
code over and over again on each project.


Releases
--------

  <http://libbsd.freedesktop.org/releases/>


Mailing List
------------

The subscription interface and web archives can be found at:

  <http://lists.freedesktop.org/mailman/listinfo/libbsd>

The mail address is:

  libbsd@lists.freedesktop.org


Source Repository
-----------------

  <http://gitweb.freedesktop.org/?p=libbsd.git>
  <git://anongit.freedesktop.org/git/libbsd>



ChangeLog
2008-03-16  Guillem Jover  <guillem@debian.org>

	* include/vis.h: Remove UC Berkeley advertising clause, as per
	<ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.
	* src/vis.c: Likewise.
	* src/unvis.c: Likewise.
	* src/heapsort.c: Likewise.
	* src/linkaddr.c: Likewise.
	* man/fgetln.3: Likewise.

2007-12-18  Petr Salinger  <Petr.Salinger@seznam.cz>

	* include/bsd/queue.h: sync with current glibc variant on GNU/kFreeBSD

2006-11-08  Aurelien Jarno  <aurelien@aurel32.net>

	* src/link_addr.c: New file.
	* include/bsd/if_dl.h: Add link_addr.c declarations.
	* Makefile (LIB_SRCS): Add link_addr.c.
	* Versions: add link_addr and link_ntoa

2006-08-25  Petr Salinger  <Petr.Salinger@seznam.cz>

	* src/humanize_number.c: New file.
	* include/libutil.h: New file with humanize_number declarations.
	* Makefile (LIB_SRCS): Add humanize_number.c.
	* Versions: Add humanize_number.

2006-03-29  Robert Millan  <rmh@aybabtu.com>

	* src/heapsort.c: New file.
	* include/bsd/stdlib.h: Add heapsort declaration.
	* Makefile (LIB_SRCS): Add heapsort.c.
	* Versions: Add heapsort.

2006-03-19  Robert Millan  <rmh@aybabtu.com>

	* src/vis.c: New file.
	* src/unvis.c: Ditto.
	* include/vis.h: Add vis/unvis declarations.
	* Makefile (LIB_SRCS): Add vis.c/unvis.c.
	* Versions: Add vis/unvis functions.

2006-03-17  Robert Millan  <rmh@aybabtu.com>

	* src/progname.c: New file.
	* include/bsd/stdlib.h: Add {get,set}progname declarations.
	* Makefile (LIB_SRCS): Add 'progname.c'.
	* Versions: Add getprogname and setprogname.

2006-02-15  Guillem Jover  <guillem@debian.org>

	* include/bsd/cdefs.h: Disable __unused, it conflicts with a struct
	member in a Linux header.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/stdlib.h: Remove <sys/types.h> and <stdio.h>.
	* include/bsd/getopt.h: Protect header from multiple inclusion.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h: Remove commented out macros.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h (setproctitle, __dead2, __unused, __printflike)
	(__FBSDID): Moved to ...
	* include/bsd/cdefs.h: ... here.
	(__unused, __printflike) [__GNUC__]: Use proper __attribute__.
	* Makefile (LIB_INCLUDES): Add 'cdefs.h'.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* arc4random.c: Move to ...
	* src/arc4random.c: ... here.
	* bsd_getopt.c: Move to ...
	* src/bsd_getopt.c:: ... here.
	* err.c: Move to ...
	* src/err.c:: ... here.
	* fgetln.c: Move to ...
	* src/fgetln.c:: ... here.
	* fmtcheck.c: Move to ...
	* src/fmtcheck.c:: ... here.
	* inet_net_pton.c: Move to ...
	* src/inet_net_pton.c:: ... here.
	* strlcat.c: Move to ...
	* src/strlcat.c:: ... here.
	* strlcpy.c: Move to ...
	* src/strlcpy.c:: ... here.
	* md5c.c: Move to ...
	* src/md5c.c:: ... here.
	* md5.copyright: Move to ...
	* src/md5.copyright:: ... here.
	* Makefile (LIB_SRCS): Automatically prefix the src/ dir.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_SRCS): Line wrap.
	(LIB_INCLUDES): List only the files, automatically prefix the dir.
	(LIB_MANS): Likewise.

2006-02-13  Robert Millan  <rmh@aybabtu.com>

	* bsd_getopt.c: New file.
	* include/bsd/getopt.h: Likewise.
	* Makefile (LIB_SRCS): Add 'bsd_getopt.c'.
	(LIB_INCLUDES): Add 'include/bsd/getopt.h'.
	* Versions: Added bsd_getopt and optreset.

2006-02-10  Robert Millan  <rmh@aybabtu.com>

	* err.c: New file.
	* include/bsd/err.h: Likewise.
	* Makefile (LIB_SRCS): Add 'err.c'.
	(LIB_INCLUDES): Add 'include/bsd/err.h'.
	* Versions: Added errc, warnc, verrc and vwarn.

2005-12-19  Aurelien Jarno <aurel32@debian.org>

	* man/arc4random.3: New file.
	* man/fgetln.3: Likewise.
	* man/fmtcheck.3: Likewise.
	* man/strlcpy.3: Likewise.
	* fmtcheck.c:: Likewise.
	* include/stdlib.h: Added fmtcheck declaration.
	* Versions: Added fmtcheck.

2005-11-01  Guillem Jover  <guillem@debian.org>

	* ip_icmp.h: Move to ...
	* include/bsd/ip_icmp.h: ... here.
	* md5.h: Move to ...
	* include/bsd/md5.h: ... here.
	* bsd.h: Move to ...
	* include/bsd/bsd.h: ... here. Split contents ...
	* include/bsd/random.h: ... here ...
	* include/bsd/queue.h: ... here ...
	* include/bsd/string.h: ... and here.
	* md5c.c: Change md5.h include line.
	* Makefile (MK_CFLAGS): Include from the proper dir.

2005-08-03  Hector Garcia  <hector@debian.org>

	* Makefile: Fixed to place soft links instead in soname, and
	shared_so instead of copying files.
	* md5.h (MD5End, MD5File, MD5FileChunk, MD5Data): Commented out
	functions not available on 'md5c.c'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_INCLUDES): Add 'md5.h'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Versions (LIBBSD_0.0): Add MD5Init, MD5Update, MD5Pad and
	MD5Final.

2005-07-26  Hector Garcia  <hector@debian.org>

	* Makefile (LIB_SRCS): Add 'md5c.c'.
	* md5c.c: Changed include to 'md5.h'.
	* md5.h: Replaced with the system header 'sys/md5.h'.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* fgetln.c (fgetln): Reindent, remove commented code and translate
	variable names to english.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* Versions: New file.
	* Makefile: Add versioned symbols support.



src/md5c.c
/*
 * MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 *
 * Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
 * rights reserved.
 *
 * License to copy and use this software is granted provided that it
 * is identified as the "RSA Data Security, Inc. MD5 Message-Digest
 * Algorithm" in all material mentioning or referencing this software
 * or this function.
 *
 * License is also granted to make and use derivative works provided
 * that such works are identified as "derived from the RSA Data
 * Security, Inc. MD5 Message-Digest Algorithm" in all material
 * mentioning or referencing the derived work.
 *
 * RSA Data Security, Inc. makes no representations concerning either
 * the merchantability of this software or the suitability of this
 * software for any particular purpose. It is provided "as is"
 * without express or implied warranty of any kind.
 *
 * These notices must be retained in any copies of any part of this
 * documentation and/or software.
 *
 * This code is the same as the code published by RSA Inc.  It has been
 * edited for clarity and style only.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libmd/md5c.c,v 1.16 2003/06/05 13:17:32 markm Exp $");

#include <sys/types.h>

#ifdef _KERNEL
#include <sys/systm.h>
#else
#include <string.h>
#endif

/*
#include <machine/endian.h>
#include <sys/endian.h>
*/
#include <bsd/md5.h>

static void MD5Transform(u_int32_t [4], const unsigned char [64]);

#ifdef _KERNEL
#define memset(x,y,z)	bzero(x,z);
#define memcpy(x,y,z)	bcopy(y, x, z)
#endif

#if (BYTE_ORDER == LITTLE_ENDIAN)
#define Encode memcpy
#define Decode memcpy
#else 

/*
 * Encodes input (u_int32_t) into output (unsigned char). Assumes len is
 * a multiple of 4.
 */

static void
Encode (unsigned char *output, u_int32_t *input, unsigned int len)
{
	unsigned int i;
	u_int32_t *op = (u_int32_t *)output;

	for (i = 0; i < len / 4; i++)
		op[i] = htole32(input[i]);
}

/*
 * Decodes input (unsigned char) into output (u_int32_t). Assumes len is
 * a multiple of 4.
 */

static void
Decode (u_int32_t *output, const unsigned char *input, unsigned int len)
{
	unsigned int i;
	const u_int32_t *ip = (const u_int32_t *)input;

	for (i = 0; i < len / 4; i++)
		output[i] = le32toh(ip[i]);
}
#endif

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions. */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits. */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/*
 * FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
 * Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
	(a) += F ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define GG(a, b, c, d, x, s, ac) { \
	(a) += G ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define HH(a, b, c, d, x, s, ac) { \
	(a) += H ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define II(a, b, c, d, x, s, ac) { \
	(a) += I ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}

/* MD5 initialization. Begins an MD5 operation, writing a new context. */

void
MD5Init (context)
	MD5_CTX *context;
{

	context->count[0] = context->count[1] = 0;

	/* Load magic initialization constants.  */
	context->state[0] = 0x67452301;
	context->state[1] = 0xefcdab89;
	context->state[2] = 0x98badcfe;
	context->state[3] = 0x10325476;
}

/* 
 * MD5 block update operation. Continues an MD5 message-digest
 * operation, processing another message block, and updating the
 * context.
 */

void
MD5Update (context, input, inputLen)
	MD5_CTX *context;
	const unsigned char *input;
	unsigned int inputLen;
{
	unsigned int i, idx, partLen;

	/* Compute number of bytes mod 64 */
	idx = (unsigned int)((context->count[0] >> 3) & 0x3F);

	/* Update number of bits */
	if ((context->count[0] += ((u_int32_t)inputLen << 3))
	    < ((u_int32_t)inputLen << 3))
		context->count[1]++;
	context->count[1] += ((u_int32_t)inputLen >> 29);

	partLen = 64 - idx;

	/* Transform as many times as possible. */
	if (inputLen >= partLen) {
		memcpy((void *)&context->buffer[idx], (const void *)input,
		    partLen);
		MD5Transform (context->state, context->buffer);

		for (i = partLen; i + 63 < inputLen; i += 64)
			MD5Transform (context->state, &input[i]);

		idx = 0;
	}
	else
		i = 0;

	/* Buffer remaining input */
	memcpy ((void *)&context->buffer[idx], (const void *)&input[i],
	    inputLen-i);
}

/*
 * MD5 padding. Adds padding followed by original length.
 */

void
MD5Pad (context)
	MD5_CTX *context;
{
	unsigned char bits[8];
	unsigned int idx, padLen;

	/* Save number of bits */
	Encode (bits, context->count, 8);

	/* Pad out to 56 mod 64. */
	idx = (unsigned int)((context->count[0] >> 3) & 0x3f);
	padLen = (idx < 56) ? (56 - idx) : (120 - idx);
	MD5Update (context, PADDING, padLen);

	/* Append length (before padding) */
	MD5Update (context, bits, 8);
}

/*
 * MD5 finalization. Ends an MD5 message-digest operation, writing the
 * the message digest and zeroizing the context.
 */

void
MD5Final (digest, context)
	unsigned char digest[16];
	MD5_CTX *context;
{
	/* Do padding. */
	MD5Pad (context);

	/* Store state in digest */
	Encode (digest, context->state, 16);

	/* Zeroize sensitive information. */
	memset ((void *)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block. */

static void
MD5Transform (state, block)
	u_int32_t state[4];
	const unsigned char block[64];
{
	u_int32_t a = state[0], b = state[1], c = state[2], d = state[3], x[16];

	Decode (x, block, 64);

	/* Round 1 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
	FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	/* Round 2 */
#define S21 5
#define S22 9
#define S23 14
#define S24 20
	GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	/* Round 3 */
#define S31 4
#define S32 11
#define S33 16
#define S34 23
	HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	/* Round 4 */
#define S41 6
#define S42 10
#define S43 15
#define S44 21
	II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	/* Zeroize sensitive information. */
	memset ((void *)x, 0, sizeof (x));
}
